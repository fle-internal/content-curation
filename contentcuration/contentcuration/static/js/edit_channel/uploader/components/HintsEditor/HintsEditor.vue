<template>
  <div>
    <div class="grey--text text--darken-1 mb-3">
      Hints
    </div>

    <div v-if="!hints || !hints.length">
      No hints yet
    </div>

    <VCard
      v-for="(hint, hintIdx) in hints"
      :key="hintIdx"
      :style="cardStyle(hintIdx)"
      flat
      data-test="hint"
      @click="onHintClick($event, hintIdx)"
    >
      <VCardText>
        <!-- eslint-disable-next-line -->
        <VLayout align-top justify-space-between>
          <VFlex xs1>
            {{ hint.order }}
          </VFlex>

          <VFlex xs4 md6 lg7>
            <p v-if="!isHintOpen(hintIdx)">
              {{ hint.hint }}
            </p>

            <keep-alive :max="5">
              <MarkdownEditor
                v-if="isHintOpen(hintIdx)"
                :markdown="hint.hint"
                @update="updateHintText($event, hintIdx)"
              />
            </keep-alive>
          </VFlex>

          <VSpacer />

          <AssessmentItemToolbar
            :displayMenu="false"
            :displayEditIcon="false"
            :canMoveUp="!isHintFirst(hintIdx)"
            :canMoveDown="!isHintLast(hintIdx)"
            class="toolbar"
            @click="onToolbarClick($event, hintIdx)"
          />
        </VLayout>
      </VCardText>
    </VCard>

    <VBtn
      flat
      color="primary"
      class="mt-3 ml-0"
      data-test="newHintBtn"
      @click="addNewHint"
    >
      New hint
    </VBtn>
  </div>
</template>

<script>

  import { AssessmentItemToolbarActions } from '../../constants';
  import { swapElements, sanitizeAssessmentItemHints } from '../../utils';

  import AssessmentItemToolbar from '../AssessmentItemToolbar/AssessmentItemToolbar.vue';
  import MarkdownEditor from '../MarkdownEditor/MarkdownEditor.vue';

  export default {
    name: 'HintsEditor',
    components: {
      AssessmentItemToolbar,
      MarkdownEditor,
    },
    model: {
      prop: 'hints',
      event: 'update',
    },
    props: {
      hints: {
        type: Array,
      },
      openHintIdx: {
        type: Number,
      },
    },
    methods: {
      emitOpen(hintIdx) {
        this.$emit('open', hintIdx);
      },
      emitClose() {
        this.$emit('close');
      },
      emitUpdate(updatedHints) {
        this.$emit('update', updatedHints);
      },
      isHintOpen(hintIdx) {
        return hintIdx === this.openHintIdx;
      },
      isHintFirst(hintIdx) {
        return hintIdx === 0;
      },
      isHintLast(hintIdx) {
        return hintIdx === this.hints.length - 1;
      },
      cardStyle(hintIdx) {
        const style = {
          border: '1px #d2d2d2 solid',
          marginTop: '-1px',
        };

        if (!this.isHintOpen(hintIdx)) {
          style.cursor = 'pointer';
        }

        return style;
      },
      moveHintUp(hintIdx) {
        if (this.isHintFirst(hintIdx)) {
          return;
        }

        let updatedHints = swapElements(this.hints, hintIdx, hintIdx - 1);
        updatedHints = updatedHints.map((hint, hintIdx) => {
          return {
            ...hint,
            order: hintIdx + 1,
          };
        });

        this.emitUpdate(updatedHints);

        if (this.isHintOpen(hintIdx)) {
          this.emitOpen(hintIdx - 1);
        } else if (this.isHintOpen(hintIdx - 1)) {
          this.emitOpen(hintIdx);
        }
      },
      moveHintDown(hintIdx) {
        if (this.isHintLast(hintIdx)) {
          return;
        }

        let updatedHints = swapElements(this.hints, hintIdx, hintIdx + 1);
        updatedHints = updatedHints.map((hint, hintIdx) => {
          return {
            ...hint,
            order: hintIdx + 1,
          };
        });

        this.emitUpdate(updatedHints);

        if (this.isHintOpen(hintIdx)) {
          this.emitOpen(hintIdx + 1);
        } else if (this.isHintOpen(hintIdx + 1)) {
          this.emitOpen(hintIdx);
        }
      },
      deleteHint(hintIdx) {
        let updatedHints = JSON.parse(JSON.stringify(this.hints));

        updatedHints.splice(hintIdx, 1);
        updatedHints = updatedHints.map((hint, hintIdx) => {
          return {
            ...hint,
            order: hintIdx + 1,
          };
        });

        this.emitUpdate(updatedHints);

        if (this.isHintOpen(hintIdx)) {
          this.emitClose();
        } else if (this.openHintIdx > hintIdx) {
          this.emitOpen(this.openHintIdx - 1);
        }
      },
      onHintClick(event, hintIdx) {
        if (this.isHintOpen(hintIdx)) {
          return;
        }

        // do not open hint on toolbar click
        if (event.target.closest('.toolbar') !== null) {
          return;
        }

        this.emitOpen(hintIdx);
      },
      onToolbarClick(action, hintIdx) {
        switch (action) {
          case AssessmentItemToolbarActions.MOVE_ITEM_UP:
            this.moveHintUp(hintIdx);
            break;

          case AssessmentItemToolbarActions.MOVE_ITEM_DOWN:
            this.moveHintDown(hintIdx);
            break;

          case AssessmentItemToolbarActions.DELETE_ITEM:
            this.deleteHint(hintIdx);
            break;
        }
      },
      addNewHint() {
        // primarily to disable adding more empty hints
        const updatedHints = sanitizeAssessmentItemHints(this.hints, true);

        updatedHints.push({
          hint: '',
          order: updatedHints.length + 1,
        });

        this.emitUpdate(updatedHints);
        this.emitOpen(updatedHints.length - 1);
      },
      updateHintText(newHintText, hintIdx) {
        if (newHintText === this.hints[hintIdx].hint) {
          return;
        }

        const updatedHints = [...this.hints];
        updatedHints[hintIdx].hint = newHintText;

        this.emitUpdate(updatedHints);
      },
    },
  };

</script>
