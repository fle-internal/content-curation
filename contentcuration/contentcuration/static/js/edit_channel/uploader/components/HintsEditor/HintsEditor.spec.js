import { shallowMount, mount } from '@vue/test-utils';

import HintsEditor from './HintsEditor';

const clickNewHintBtn = wrapper => {
  wrapper
    .find('[data-test=newHintBtn]')
    .find('button')
    .trigger('click');
};

const clickHint = (wrapper, hintIdx) => {
  wrapper
    .findAll('[data-test=hint]')
    .at(hintIdx)
    .trigger('click');
};

const clickMoveHintUp = (wrapper, hintIdx) => {
  wrapper
    .findAll('[data-test=toolbarIconArrowUp]')
    .at(hintIdx)
    .trigger('click');
};

const clickMoveHintDown = (wrapper, hintIdx) => {
  wrapper
    .findAll('[data-test=toolbarIconArrowDown]')
    .at(hintIdx)
    .trigger('click');
};

const clickDeleteHint = (wrapper, hintIdx) => {
  wrapper
    .findAll('[data-test=toolbarIconDelete]')
    .at(hintIdx)
    .trigger('click');
};

const updateOpenHintText = (wrapper, newHintText) => {
  // only one hint can be open
  wrapper.find('[data-test=editHintTextInput]').setValue(newHintText);
};

describe('HintsEditor', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = mount(HintsEditor, {
      propsData: {
        hints: [{ hint: 'First hint', order: 1 }, { hint: 'Second hint', order: 2 }],
      },
    });
  });

  it('smoke test', () => {
    const wrapper = shallowMount(HintsEditor);

    expect(wrapper.isVueInstance()).toBe(true);
  });

  describe('with no hints', () => {
    beforeEach(() => {
      wrapper.setProps({
        hints: [],
      });
    });

    it('renders a placeholder', () => {
      expect(wrapper.html()).toContain('No hints yet');
    });
  });

  it('renders all hints', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe('on new hint button click', () => {
    beforeEach(() => {
      clickNewHintBtn(wrapper);
    });

    it('emits update event with a payload containing all hints + new hint', () => {
      expect(wrapper.emitted().update).toBeTruthy();
      expect(wrapper.emitted().update.length).toBe(1);
      expect(wrapper.emitted().update[0][0]).toEqual([
        { hint: 'First hint', order: 1 },
        { hint: 'Second hint', order: 2 },
        { hint: '', order: 3 },
      ]);
    });

    it('emits open event with a new hint idx', () => {
      expect(wrapper.emitted().open).toBeTruthy();
      expect(wrapper.emitted().open.length).toBe(1);
      expect(wrapper.emitted().open[0][0]).toBe(2);
    });
  });

  describe('on hint click', () => {
    it('emits open event with a correct hint idx', () => {
      clickHint(wrapper, 1);

      expect(wrapper.emitted().open).toBeTruthy();
      expect(wrapper.emitted().open.length).toBe(1);
      expect(wrapper.emitted().open[0][0]).toBe(1);
    });
  });

  describe('on hint text update', () => {
    beforeEach(() => {
      wrapper = mount(HintsEditor, {
        propsData: {
          hints: [{ hint: 'First hint', order: 1 }, { hint: 'Second hint', order: 2 }],
          openHintIdx: 1,
        },
      });

      updateOpenHintText(wrapper, 'Second updated hint');
    });

    it('emits update event with a payload containing updated hints', () => {
      expect(wrapper.emitted().update).toBeTruthy();
      expect(wrapper.emitted().update.length).toBe(1);
      expect(wrapper.emitted().update[0][0]).toEqual([
        { hint: 'First hint', order: 1 },
        { hint: 'Second updated hint', order: 2 },
      ]);
    });
  });

  describe('on move hint up click', () => {
    it('emits update event with a payload containing updated and properly ordered hints', () => {
      clickMoveHintUp(wrapper, 1);

      expect(wrapper.emitted().update).toBeTruthy();
      expect(wrapper.emitted().update.length).toBe(1);
      expect(wrapper.emitted().update[0][0]).toEqual([
        { hint: 'Second hint', order: 1 },
        { hint: 'First hint', order: 2 },
      ]);
    });

    describe('if moved hint was open', () => {
      beforeEach(() => {
        wrapper.setProps({
          openHintIdx: 1,
        });

        clickMoveHintUp(wrapper, 1);
      });

      it('emits open event with updated hint index', () => {
        expect(wrapper.emitted().open).toBeTruthy();
        expect(wrapper.emitted().open.length).toBe(1);
        expect(wrapper.emitted().open[0][0]).toBe(0);
      });
    });

    describe('if a hint above a moved hint was open', () => {
      beforeEach(() => {
        wrapper.setProps({
          openHintIdx: 0,
        });

        clickMoveHintUp(wrapper, 1);
      });

      it('emits open event with updated, originally open, hint index', () => {
        expect(wrapper.emitted().open).toBeTruthy();
        expect(wrapper.emitted().open.length).toBe(1);
        expect(wrapper.emitted().open[0][0]).toBe(1);
      });
    });
  });

  describe('on move hint down click', () => {
    it('emits update event with a payload containing updated and properly ordered hints', () => {
      clickMoveHintDown(wrapper, 0);

      expect(wrapper.emitted().update).toBeTruthy();
      expect(wrapper.emitted().update.length).toBe(1);
      expect(wrapper.emitted().update[0][0]).toEqual([
        { hint: 'Second hint', order: 1 },
        { hint: 'First hint', order: 2 },
      ]);
    });

    describe('if moved hint was open', () => {
      beforeEach(() => {
        wrapper.setProps({
          openHintIdx: 0,
        });

        clickMoveHintDown(wrapper, 0);
      });

      it('emits open event with updated hint index', () => {
        expect(wrapper.emitted().open).toBeTruthy();
        expect(wrapper.emitted().open.length).toBe(1);
        expect(wrapper.emitted().open[0][0]).toBe(1);
      });
    });

    describe('if a hint below a moved hint was open', () => {
      beforeEach(() => {
        wrapper.setProps({
          openHintIdx: 1,
        });

        clickMoveHintDown(wrapper, 0);
      });

      it('emits open event with updated, originally open, hint index', () => {
        expect(wrapper.emitted().open).toBeTruthy();
        expect(wrapper.emitted().open.length).toBe(1);
        expect(wrapper.emitted().open[0][0]).toBe(0);
      });
    });
  });

  describe('on delete hint click', () => {
    it('emits update event with a payload containing updated and properly ordered hints', () => {
      clickDeleteHint(wrapper, 0);

      expect(wrapper.emitted().update).toBeTruthy();
      expect(wrapper.emitted().update.length).toBe(1);
      expect(wrapper.emitted().update[0][0]).toEqual([{ hint: 'Second hint', order: 1 }]);
    });

    describe('if deleted hint was open', () => {
      beforeEach(() => {
        wrapper.setProps({
          openHintIdx: 0,
        });

        clickDeleteHint(wrapper, 0);
      });

      it('emits close event', () => {
        expect(wrapper.emitted().close).toBeTruthy();
        expect(wrapper.emitted().close.length).toBe(1);
      });
    });

    describe('if a hint below a deleted hint was open', () => {
      beforeEach(() => {
        wrapper.setProps({
          openHintIdx: 1,
        });

        clickDeleteHint(wrapper, 0);
      });

      it('emits open event with updated, originally open, hint index', () => {
        expect(wrapper.emitted().open).toBeTruthy();
        expect(wrapper.emitted().open.length).toBe(1);
        expect(wrapper.emitted().open[0][0]).toBe(0);
      });
    });
  });
});
