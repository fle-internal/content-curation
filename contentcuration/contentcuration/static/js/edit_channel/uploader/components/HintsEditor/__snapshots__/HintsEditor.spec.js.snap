// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HintsEditor renders all hints 1`] = `
<div>
  <div class="grey--text text--darken-1 mb-3">
    Hints
  </div>
  <!---->
  <div data-test="hint" class="card-border-light">
    <div class="v-card v-card--flat v-sheet theme--light hint closed">
      <div class="v-card__text">
        <div class="layout align-top">
          <div class="flex xs1">
            1
          </div>
          <div class="flex"><span name="fade">
                  First hint
                </span></div>
          <div class="spacer"></div>
          <div class="flex">
            <div class="layout justify-end toolbar">
              <div class="flex toolbar-item"><span class="v-tooltip v-tooltip--top"><div class="v-tooltip__content" style="left: 0px; max-width: auto; opacity: 0; top: 12px; z-index: 0; display: none;"> <span>Move up</span></div><button disabled="disabled" type="button" class="v-btn v-btn--disabled v-btn--icon theme--light" data-test="toolbarIcon-MOVE_ITEM_UP">
                <div class="v-btn__content"><i aria-hidden="true" class="v-icon material-icons theme--light grey--text text--lighten-2">keyboard_arrow_up</i></div>
              </button></span>
            </div>
            <div class="flex toolbar-item"><span class="v-tooltip v-tooltip--top"><div class="v-tooltip__content" style="left: 0px; max-width: auto; opacity: 0; top: 12px; z-index: 0; display: none;"> <span>Move down</span></div><button type="button" class="v-btn v-btn--icon theme--light" data-test="toolbarIcon-MOVE_ITEM_DOWN">
              <div class="v-btn__content"><i aria-hidden="true" class="v-icon material-icons theme--light grey--text text--darken-1">keyboard_arrow_down</i></div>
            </button></span>
          </div>
          <div class="flex toolbar-item"><span class="v-tooltip v-tooltip--top"><div class="v-tooltip__content" style="left: 0px; max-width: auto; opacity: 0; top: 12px; z-index: 0; display: none;"> <span>Delete</span></div><button type="button" class="v-btn v-btn--icon theme--light" data-test="toolbarIcon-DELETE_ITEM">
            <div class="v-btn__content"><i aria-hidden="true" class="v-icon material-icons theme--light grey--text text--darken-1">close</i></div>
          </button></span>
        </div>
        <!---->
      </div>
    </div>
  </div>
</div>
</div>
</div>
<div data-test="hint" class="card-border-light">
  <div class="v-card v-card--flat v-sheet theme--light hint closed">
    <div class="v-card__text">
      <div class="layout align-top">
        <div class="flex xs1">
          2
        </div>
        <div class="flex"><span name="fade">
                  Second hint
                </span></div>
        <div class="spacer"></div>
        <div class="flex">
          <div class="layout justify-end toolbar">
            <div class="flex toolbar-item"><span class="v-tooltip v-tooltip--top"><div class="v-tooltip__content" style="left: 0px; max-width: auto; opacity: 0; top: 12px; z-index: 0; display: none;"> <span>Move up</span></div><button type="button" class="v-btn v-btn--icon theme--light" data-test="toolbarIcon-MOVE_ITEM_UP">
              <div class="v-btn__content"><i aria-hidden="true" class="v-icon material-icons theme--light grey--text text--darken-1">keyboard_arrow_up</i></div>
            </button></span>
          </div>
          <div class="flex toolbar-item"><span class="v-tooltip v-tooltip--top"><div class="v-tooltip__content" style="left: 0px; max-width: auto; opacity: 0; top: 12px; z-index: 0; display: none;"> <span>Move down</span></div><button disabled="disabled" type="button" class="v-btn v-btn--disabled v-btn--icon theme--light" data-test="toolbarIcon-MOVE_ITEM_DOWN">
            <div class="v-btn__content"><i aria-hidden="true" class="v-icon material-icons theme--light grey--text text--lighten-2">keyboard_arrow_down</i></div>
          </button></span>
        </div>
        <div class="flex toolbar-item"><span class="v-tooltip v-tooltip--top"><div class="v-tooltip__content" style="left: 0px; max-width: auto; opacity: 0; top: 12px; z-index: 0; display: none;"> <span>Delete</span></div><button type="button" class="v-btn v-btn--icon theme--light" data-test="toolbarIcon-DELETE_ITEM">
          <div class="v-btn__content"><i aria-hidden="true" class="v-icon material-icons theme--light grey--text text--darken-1">close</i></div>
        </button></span>
      </div>
      <!---->
    </div>
  </div>
</div>
</div>
</div>
</div> <button type="button" class="mt-3 ml-0 v-btn v-btn--flat theme--light primary--text" data-test="newHintBtn">
  <div class="v-btn__content">
    New hint
  </div>
</button></div>
`;
